gpushift/sklearn/meanshift.py

```diff
@@ -0,0 +1,321 @@
+        #points_j = LazyTensor(points[:,:,None,:]) # (B,N,1,C)
+
+        #pairwise_distances_ij = self.distance_metric(points_i, points_j) # (B,N,N,1)
+
+        #neighbours_ij = (torch.Tensor(self.bandwidth**2) - pairwise_distances_ij).step() # (B,N,N,1) 0 or 1
+
+        #intensities = self.kernel(pairwise_distances_ij).sum(2).unsqueeze(-1) # (B,N,1,1)
+        #possible_clusters = LazyTensor(-intensities).argKmin(self.max_clusters, dim=1) # (B,1,K)
+
+        #are_cluster_centers_ = torch.ones(B,N).bool()
+        #for b in range(B):
+        #    possible_cluster_indices = possible_clusters[b,0] # (K,)
+        #    possible_cluster_intensities = intensities[b, possible_cluster_indices,0,0] # (K,)
+
+        #    argsort_res = torch.argsort(possible_cluster_intensities) # Ascending
+        #    sorted_cluster_indices = possible_cluster_indices[argsort_res] # (K,)
+
+        #    for idx in sorted_cluster_indices:
+        #        import pdb; pdb.set_trace()
+        #        are_neighbours = neighbours_ij[b,idx] # (N,1)
+        #        are_cluster_centers_[b,are_neighbours] = False
+        #        are_cluster_centers_[b,idx] = True
+
+        return cluster_centers
+
+class MeanShift(BaseEstimator, ClusterMixin):
+    r"""
+    This class implements the sklean interface for a gpu-accelerated mean 
+    shift algrithm.
+    """
+
+    def __init__(self, bandwidth, distance_metric='composite', kernel='gaussian', 
+                       n_iter=15, max_clusters=500, blurring=False, use_keops=True):
+        r"""
+        Inputs:
+            :param bandwidth: int, bandwidth of the kernel. Some kernels such
+                as 'gaussian' have infinite support, so this parameter becomes
+                the sigma of the curve. For other kernels such as 'flat' and 
+                'epanechnikov', it sets the support of the kernel.
+
+            :param distance_metric: str
+
+            :param kernel: str, one of pre-defined values which controls which
+                kernel the algorithm will use.
+
+            :param n_iter: int
+
+            :param max_clusters: int
+
+            :param blurring: bool, whether to run the meanshift algorithm in a
+                blurring way (each point moves and the KDE landscape changes 
+                at each iteration), or non-blurring (keep the KDE landscape
+                constant, and move a set of reference points up the gradient
+                instead).
+
+            :param use_keops: bool, whether to use the gpu-accelerated version
+                of mean shift.
+        """
+        self.bandwidth = bandwidth
+        self.distance_metric = distance_metric
+        self.kernel = kernel
+        self.n_iter = n_iter
+        self.max_clusters = max_clusters
+        self.blurring = blurring
+        self.use_keops = use_keops
+
+        self.cluster_centers_ = None
+
+        self.meanshift_step = MeanShiftStep(bandwidth=bandwidth, kernel=kernel, use_keops=use_keops)
```

Comment:
Was it your intention to not pass the distance metric here? I thought not only the final clustering but also every MS step should use the chosen metric.